File: CS/MultipleDataSources/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.Xpo;
using DevExpress.DataAccess.Excel;

namespace MultipleDataSources {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();

            // Create XPO data sources with one-to-many relationship.
            Session session = new Session();

            var categories = new CustomXPCollection<Category>(session);
            new Category(session) { CategoryName = "Beverages" }.Save();
            new Category(session) { CategoryName = "Condiments" }.Save();
            new Category(session) { CategoryName = "Seafood" }.Save();

            var products = new CustomXPCollection<Product>(session);
            new Product(session) { ProductName = "Ikura", Category = categories.First(x => x.CategoryName == "Seafood") }.Save();
            new Product(session) { ProductName = "Ipoh Coffee", Category = categories.First(x => x.CategoryName == "Beverages") }.Save();
            new Product(session) { ProductName = "Aniseed Syrup", Category = categories.First(x => x.CategoryName == "Condiments") }.Save();

            //Create a data set and load a data table from an XML file.
            DataSet dsContacts = new DataSet();
            dsContacts.ReadXml("Contacts.xml");

            // Create an Excel data source.
            ExcelDataSource myExcelSource = new ExcelDataSource();
            myExcelSource.FileName = "OrderDetails.xlsx";
            ExcelWorksheetSettings worksheetSettings = new ExcelWorksheetSettings("Sales report", "B2:H18");
            myExcelSource.SourceOptions = new ExcelSourceOptions(worksheetSettings);
            myExcelSource.SourceOptions.UseFirstRowAsHeader = true;
            myExcelSource.Fill();

            #region #SetUseForMailMerge
            // Set the Application data source which cannot be edited or removed by the end-user.
            this.snapControl1.DataSource = dsContacts.Tables[0];
            // Add Document data sources which persist for the currently loaded document only.
            this.snapControl1.Document.BeginUpdateDataSource();
            this.snapControl1.Document.DataSources.Add("Products", products);
            this.snapControl1.Document.DataSources.Add("Categories", categories);
            this.snapControl1.Document.DataSources.Add("Sales", myExcelSource);
            this.snapControl1.Document.EndUpdateDataSource();

            // Set the default (Application) data source as the mail merge source ("Use for Mail Merge" option).
            this.snapControl1.Options.SnapMailMergeVisualOptions.DataSourceName = "";
            // Uncomment the following line to set the Sales data source as the mail merge source.
            // this.snapControl1.Options.SnapMailMergeVisualOptions.DataSourceName = "Sales";
            #endregion #SetUseForMailMerge
        }
    }
}

File: VB/MultipleDataSources/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.Xpo
Imports DevExpress.DataAccess.Excel

Namespace MultipleDataSources
    Partial Public Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()

            ' Create XPO data sources with one-to-many relationship.
            Dim session As New Session()

            Dim categories = New CustomXPCollection(Of Category)(session)
            CType(New Category(session) With {.CategoryName = "Beverages"}, Category).Save()
            CType(New Category(session) With {.CategoryName = "Condiments"}, Category).Save()
            CType(New Category(session) With {.CategoryName = "Seafood"}, Category).Save()

            Dim products = New CustomXPCollection(Of Product)(session)
            CType(New Product(session) With {.ProductName = "Ikura", .Category = categories.First(Function(x) x.CategoryName = "Seafood")}, Product).Save()
            CType(New Product(session) With {.ProductName = "Ipoh Coffee", .Category = categories.First(Function(x) x.CategoryName = "Beverages")}, Product).Save()
            CType(New Product(session) With {.ProductName = "Aniseed Syrup", .Category = categories.First(Function(x) x.CategoryName = "Condiments")}, Product).Save()

            'Create a data set and load a data table from an XML file.
            Dim dsContacts As New DataSet()
            dsContacts.ReadXml("Contacts.xml")

            ' Create an Excel data source.
            Dim myExcelSource As New ExcelDataSource()
            myExcelSource.FileName = "OrderDetails.xlsx"
            Dim worksheetSettings As New ExcelWorksheetSettings("Sales report", "B2:H18")
            myExcelSource.SourceOptions = New ExcelSourceOptions(worksheetSettings)
            myExcelSource.SourceOptions.UseFirstRowAsHeader = True
            myExcelSource.Fill()

'            #Region "#SetUseForMailMerge"
            ' Set the Application data source which cannot be edited or removed by the end-user.
            Me.snapControl1.DataSource = dsContacts.Tables(0)
            ' Add Document data sources which persist for the currently loaded document only.
            Me.snapControl1.Document.BeginUpdateDataSource()
            Me.snapControl1.Document.DataSources.Add("Products", products)
            Me.snapControl1.Document.DataSources.Add("Categories", categories)
            Me.snapControl1.Document.DataSources.Add("Sales", myExcelSource)
            Me.snapControl1.Document.EndUpdateDataSource()

            ' Set the default (Application) data source as the mail merge source ("Use for Mail Merge" option).
            Me.snapControl1.Options.SnapMailMergeVisualOptions.DataSourceName = ""
            ' Uncomment the following line to set the Sales data source as the mail merge source.
            ' this.snapControl1.Options.SnapMailMergeVisualOptions.DataSourceName = "Sales";
'            #End Region ' #SetUseForMailMerge
        End Sub
    End Class
End Namespace
